gcovr/gcov.py

```diff
@@ -231,15 +231,20 @@ def guess_source_file_name_heuristics(
         if os.path.exists(fname):
             return fname
 
+    # 4. Try using the path to the gcov file
+    fname = os.path.join(os.path.dirname(data_fname), gcovname)
+    if os.path.exists(fname):
+        return fname
+
```

Comment:
I'm quite afraid of touching any code in this part of gcovr – whether I like the heuristics or not, any change could break backwards compatibility. Unless there's a good reason, new heuristics should therefore be added at the end.

Could you explain a bit more about why this heuristic is appropriate and correct? As far as I understand, the `data_fname` is the `.gcov` file, which will be created in whatever directory that `gcov` was invoked in as per the `potential_wd` heuristics in `process_datafile()`.

For which specific test cases does the current behavior become a problem?

Reply:
I can move it after the g da part. The cause is written in the commit comment.
```
gcc-11 creates in the gcda files in a sub directory which results in an error later. The file is used as testApp/test/tmp.cpp but correct is testApp/tmp.cpp.

(DEBUG) Running gcov: '/usr/bin/gcov-11 /gcovr/gcovr/tests/shared_lib/testApp/test/a-tmp.gcda --branch-counts --branch-probabilities --demangled-names --hash-filenames --object-directory /gcovr/gcovr/tests/shared_lib/testApp/test' in '/gcovr/gcovr/tests/shared_lib/testApp'
(DEBUG) Finding source file corresponding to a gcov data file
  currdir      /gcovr/gcovr/tests/shared_lib
  gcov_fname   /gcovr/gcovr/tests/shared_lib/testApp/tmp.cpp##95fbba3ae32e8157841b28787cf829e3.gcov
  source_fname /gcovr/gcovr/tests/shared_lib/testApp/test/a-tmp.gcda
  root         /gcovr/gcovr/tests/shared_lib
  fname        /gcovr/gcovr/tests/shared_lib/testApp/test/tmp.cpp
(DEBUG) Parsing coverage data for file /gcovr/gcovr/tests/shared_lib/testApp/test/tmp.cpp
```

Previous versions created the gcda file directly in the test app folder.